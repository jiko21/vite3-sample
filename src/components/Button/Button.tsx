import { css } from '@emotion/react';
import React, { ButtonHTMLAttributes } from 'react';

export type ButtonStyle = 'primary' | 'danger';
type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> & {
  buttonStyle?: ButtonStyle;
  disabled?: boolean;
  testId?: string;
};

const baseStyle = css`
  border: none;
  border-radius: 4px;
  font-weight: 700;
  height: 32px;
  line-height: 16px;
  min-width: 60px;
  padding: 8px 4px;
  position: relative;
  :active {
    top: 1px;
  }
`;

const primaryStyle = css`
  background-color: #00c1cf;
  color: white;
  :hover {
    background-color: #5ec9d1;
    color: white;
  }
`;

const dangerStyle = css`
  background-color: #ff002b;
  color: white;
  :hover {
    background-color: #ff3b3b;
    color: white;
  }
`;

const disabledStyle = css`
  background-color: #dfdfdf;
  color: #3c3c3c;
  :active {
    top: 0;
  }
`;

export const Button: React.FC<ButtonProps> = ({
  buttonStyle = 'primary',
  disabled = false,
  children,
  testId,
  onClick,
  ...props
}) => {
  const style = disabled ? disabledStyle : buttonStyle === 'danger' ? dangerStyle : primaryStyle;
  return (
    <button
      data-testid={`${testId}-button`}
      css={[baseStyle, style]}
      onClick={(e) => {
        if (!disabled && onClick) onClick(e);
      }}
      {...props}
    >
      {children}
    </button>
  );
};
