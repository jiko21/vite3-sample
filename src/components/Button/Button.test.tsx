import { afterEach, describe, expect, it, vi } from 'vitest';
import { Button } from './Button';
import { cleanup, fireEvent, render, screen } from '@testing-library/react';

describe('components/Button', () => {
  afterEach(cleanup);

  it('correctly render', () => {
    const { container } = render(<Button>Test</Button>);
    expect(screen.getByText('Test')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('correctly fire event', () => {
    const onClickMock = vi.fn();
    render(<Button onClick={onClickMock}>Test</Button>);
    fireEvent.click(screen.getByText('Test'));
    expect(onClickMock).toBeCalled();
  });
});
