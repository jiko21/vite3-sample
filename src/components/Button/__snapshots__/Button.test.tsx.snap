// Vitest Snapshot v1

exports[`components/Button > correctly render 1`] = `
<div>
  <button
    class="css-t9php5-Button"
    data-testid="undefined-button"
  >
    Test
  </button>
</div>
`;
